diff -urN a/configure.ac b/configure.ac
--- a/configure.ac	2013-10-12 18:05:00.000000000 +0200
+++ b/configure.ac	2013-10-12 18:26:41.123055859 +0200
@@ -109,8 +109,12 @@
 	JANSSON_LIBS=-ljansson
 fi
 
-LIBCURL_CHECK_CONFIG(, 7.15.2, ,
-  [AC_MSG_ERROR([Missing required libcurl >= 7.15.2])])
+PKG_CHECK_MODULES([LIBCURL], [libcurl >= 7.18.2], ,[AC_MSG_ERROR([Missing required libcurl dev >= 7.18.2])])
+if echo "$LIBCURL_CFLAGS" | grep '@CPPFLAG_CURL_STATICLIB@' >/dev/null 2>&1; then
+	AC_MSG_WARN([Your libcurl pkgconfig file is broken, applying workaround])
+	LIBCURL_CFLAGS=`echo "$LIBCURL_CFLAGS" | sed 's/@CPPFLAG_CURL_STATICLIB@//'`
+fi
+AC_SUBST(LIBCURL_LIBS)
 
 AC_SUBST(JANSSON_LIBS)
 AC_SUBST(PTHREAD_FLAGS)
diff -urN a/Makefile.am b/Makefile.am
--- a/Makefile.am	2013-10-12 18:05:00.000000000 +0200
+++ b/Makefile.am	2013-10-12 18:23:00.258511998 +0200
@@ -26,6 +26,6 @@
 minerd_SOURCES += sha2-arm.S scrypt-arm.S
 endif
 minerd_LDFLAGS	= $(PTHREAD_FLAGS)
-minerd_LDADD	= @LIBCURL@ @JANSSON_LIBS@ @PTHREAD_LIBS@ @WS2_LIBS@
-minerd_CPPFLAGS = @LIBCURL_CPPFLAGS@
+minerd_LDADD	= -lcurl @JANSSON_LIBS@ @PTHREAD_LIBS@ @WS2_LIBS@
+#minerd_CPPFLAGS = @LIBCURL_CPPFLAGS@
 
