--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -322,6 +322,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -365,6 +367,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -406,6 +410,8 @@ static struct board_info __initdata boar
 	.has_enet0			= 1,
 
 	.enet0 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -439,6 +445,8 @@ static struct board_info __initdata boar
 	.has_enet0			= 1,
 
 	.enet0 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -543,6 +551,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -615,6 +625,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -675,6 +687,8 @@ static struct board_info __initdata boar
 	.has_pci			= 1,
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -730,6 +744,8 @@ static struct board_info __initdata boar
 	.has_pci			= 1,
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -772,6 +788,8 @@ static struct board_info __initdata boar
 	.has_pci			= 1,
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -838,6 +856,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -906,6 +926,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -968,6 +990,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1013,6 +1037,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1036,6 +1062,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1057,6 +1085,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1076,6 +1106,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1105,6 +1137,8 @@ static struct board_info __initdata boar
 	.has_uart0 			= 1,
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1220,6 +1254,8 @@ static struct board_info __initdata boar
 	},
 	.enet1 = {
 		/* it has BP_ENET_EXTERNAL_PHY */
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1289,6 +1325,8 @@ static struct board_info __initdata boar
                 .use_internal_phy       = 1,
         },
         .enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
                 .force_speed_100        = 1,
                 .force_duplex_full      = 1,
 	},
@@ -1346,6 +1384,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1379,6 +1419,8 @@ static struct board_info __initdata boar
 	.has_pci			= 1,
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1441,6 +1483,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1493,6 +1537,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1535,6 +1581,8 @@ static struct board_info __initdata boar
 	.has_pci                        = 1,
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100        = 1,
 		.force_duplex_full      = 1,
  	},
@@ -1631,6 +1679,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100        = 1,
 		.force_duplex_full      = 1,
 	},
@@ -1748,6 +1798,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1766,6 +1818,8 @@ static struct board_info __initdata boar
 	.has_pci			= 1,
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2435,6 +2489,8 @@ static struct board_info __initdata boar
 
 	.has_enet1			= 1,
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2510,6 +2566,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100        = 1,
 		.force_duplex_full      = 1,
 	},
@@ -2578,6 +2636,8 @@ static struct board_info __initdata boar
 	.has_pci		= 1,
 
 	.enet0 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.has_phy 		= 1,
 		.use_internal_phy 	= 1,
 	},
@@ -2652,6 +2712,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy                = 1,
+		.phy_id                 = 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
